File: python/P394.py
#!/usr/bin/python

class Solution(object):
	def decodeString(self, s):
		"""
		:type s: str
		:rtype: str
		"""
		self.res, self.i = [], 0
		def decodeRec(s):
			res = []
			while self.i < len(s):
				if s[self.i] == "]": 
					self.i += 1
					return res
				elif s[self.i].isalpha():
					res.append(s[self.i])
					self.i += 1
				elif s[self.i].isdigit():
					i = self.i
					while s[i].isdigit(): i += 1
					mul = int(s[self.i:i])
					self.i = i+1
					subres = decodeRec(s)
					res += mul * subres
			self.res += res
		decodeRec(s)
		return "".join(self.res)

s = Solution()
print(s.decodeString("2[abc]3[cd]ef"))

File: src/P394.cpp
#include <iostream>
#include "../include/common.h"
using namespace std;

class Solution {
	static bool isdigit(char ch) {
		return ch >= '0' && ch <= '9';
	}
	
	string decode_helper(string s, int &idx) {
		string res;
		
		while (idx < s.size() && s[idx] != ']') {
			while (idx < s.size() && !isdigit(s[idx])) {
				if (s[idx] == ']') {
					idx++;
					return res;
				}

				res.push_back(s[idx++]);
			}

			int multiplier = 0;
			cout << multiplier << endl;
			while (idx < s.size() && isdigit(s[idx])) {
				multiplier = multiplier * 10 + s[idx++] - '0';
			}

			// skip '['
			string subres = decode_helper(s, ++idx);
			for (int i = 0; i < multiplier; i++) {
				res.append(subres);
			}

			if (idx < s.size() && s[idx] == ']') {
				idx++;
			}
			
		}
		return res;
	}
	
public:
	string decodeString(string s) {
		string res;
		int i = 0;
		
		res = decode_helper(s, i);
		
		return res;
	}
};

int main(int argc, char *argv[]) {
	Solution s;
	
	cout << s.decodeString("3[z]2[2[y]pq4[2[jk]e1[f]]]ef") << endl;
	
	return 0;
}